libs/compact-map/src/Data/Compact/KeyMap.hs

```diff
@@ -321,19 +356,54 @@ delete k hm = delete' (keyPath k) k hm
 oneE :: Int -> KeyMap v -> KeyMap v
 oneE _ Empty = Empty
 oneE i x = One i x
+{-# INLINE oneE #-}
 
 -- Float Empty's up over Two
 twoE :: Bitmap -> KeyMap v -> KeyMap v -> KeyMap v
 twoE _ Empty Empty = Empty
 twoE bmap x Empty = oneE (ith bmap 0) x
 twoE bmap Empty x = oneE (ith bmap 1) x
 twoE bmap x y = Two bmap x y
+{-# INLINE twoE #-}
 
--- Float Empty's up over BitmpIndexed, Note that if the size of the arr
--- becomes 2, then rebuild with Two rather than BitmapIndexed
-bitmapE :: Bitmap -> PArray (KeyMap v) -> KeyMap v
-bitmapE bmap arr | isize arr == 2 = twoE bmap (index arr 0) (index arr 1)
-bitmapE bmap arr = bitmapIndexedOrFull bmap arr
+-- | Create a 'BitmapIndexed' or 'Full' or 'One' or 'Two' node depending on the size of 'arr'
+--   and dropping all Empty nodes.  Use this only where things can become empty (delete, intersect, etc)
+dropEmpty :: Bitmap -> PArray (KeyMap v) -> KeyMap v
+dropEmpty _ arr | isize arr == 0 = Empty
+dropEmpty b arr | isize arr == 1 = oneE (head (bitmapToList b)) (index arr 0)
```

Comment:
Partial functions like `head` should never be used, even if they are look impossible. It is not obvious that `b /= 0`.

```haskell
> head $ bitmapToList (0 :: Int)
*** Exception: Prelude.head: empty list
```

 It is very hard to find the source of these sort of errors, because all you get is this when it happens: `*** Exception: Prelude.head: empty list`, without any information where in the codebase it originated from.

Reply:
The KeyMap data structure maintains an invariant that the number of bits set in a BitMap is the same as the size of the Array. I removed the calls to 'head' and replaced them with a case analysis that calls error with a message that locates the call if this invariant is violated. I also make the invariant explicit in a few cases where it is maintained.